File: src/cycod/Configuration/ConfigValue.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class ConfigValue
{
    public ConfigValue()
    {
        _value = null;
        Source = ConfigSource.NotFound;
        IsSecret = false;
    }

    public ConfigValue(object? value)
    {
        _value = value;
        Source = ConfigSource.Default;
        IsSecret = false;
    }
    
    public ConfigValue(object? value, ConfigSource source)
    {
        _value = value;
        Source = source;
        IsSecret = false;
    }
    
    public ConfigValue(object? value, ConfigSource source, bool isSecret)
    {
        _value = value;
        Source = source;
        IsSecret = isSecret;
    }

    public object? Value => _value;
    
    /// <summary>
    /// Gets the source of this configuration value.
    /// </summary>
    public ConfigSource Source { get; }

    /// <summary>
    /// Gets the configuration file associated with this value, if any.
    /// </summary>
    public ConfigFile? File { get; internal set; }

    /// <summary>
    /// Gets whether this value should be treated as a secret.
    /// </summary>
    public bool IsSecret { get; }

    public string? AsString()
    {
        return _value?.ToString();
    }
    
    /// <summary>
    /// Returns an obfuscated version of the value if it's a secret.
    /// </summary>
    /// <returns>The obfuscated value if it's a secret, otherwise the normal string representation.</returns>
    public string? AsObfuscated()
    {
        if (!IsSecret || _value == null)
        {
            return AsString();
        }
        
        var valueStr = _value.ToString() ?? string.Empty;
        if (valueStr.Length <= 4)
        {
            return "****";
        }
        
        // Show first 2 and last 2 characters, obfuscate the rest
        return valueStr.Substring(0, 2) + new string('*', valueStr.Length - 4) + valueStr.Substring(valueStr.Length - 2);
    }

    public int AsInt(int defaultValue = 0)
    {
        if (_value is int intValue)
        {
            return intValue;
        }

        if (_value is string stringValue && int.TryParse(stringValue, out var parsedInt))
        {
            return parsedInt;
        }

        return defaultValue;
    }

    public bool AsBool(bool defaultValue = false)
    {
        if (_value is bool boolValue)
        {
            return boolValue;
        }

        if (_value is string stringValue)
        {
            if (bool.TryParse(stringValue, out var parsedBool))
            {
                return parsedBool;
            }

            return stringValue.ToLower() == "true" || stringValue == "1";
        }

        if (_value is int intValue)
        {
            return intValue != 0;
        }

        return defaultValue;
    }

    public List<string> AsList()
    {
        // Handle null value
        if (_value == null)
        {
            ConsoleHelpers.WriteDebugLine($"AsList: null value");
            return new List<string>();
        }

        // Handle List<object>
        if (_value is List<object> objList)
        {
            var list = objList.Select(o => o?.ToString() ?? string.Empty).ToList();
            ConsoleHelpers.WriteDebugLine($"AsList: List<object> [{string.Join(", ", list)}]");
            return list;
        }

        // Handle List<string>
        if (_value is List<string> stringList)
        {
            ConsoleHelpers.WriteDebugLine($"AsList: List<string> [{string.Join(", ", stringList)}]");
            return stringList;
        }

        // Handle arrays (object[] and string[])
        if (_value is object[] objArray)
        {
            var list = objArray.Select(o => o?.ToString() ?? string.Empty).ToList();
            ConsoleHelpers.WriteDebugLine($"AsList: object[] [{string.Join(", ", list)}]");
            return list;
        }

        if (_value is string[] strArray)
        {
            ConsoleHelpers.WriteDebugLine($"AsList: string[] [{string.Join(", ", strArray)}]");
            return strArray.ToList();
        }

        // Handle single string
        if (_value is string singleString)
        {
            // Check if it's an empty array representation from YAML
            if (singleString == "[]")
            {
                ConsoleHelpers.WriteDebugLine($"AsList: empty array string representation");
                return new List<string>();
            }
            
            ConsoleHelpers.WriteDebugLine($"AsList: single string {singleString}");
            return new List<string> { singleString };
        }

        // Handle IEnumerable<object> (catches more generic collections)
        if (_value is System.Collections.IEnumerable enumerable && 
            !(_value is string)) // Exclude strings which are IEnumerable<char>
        {
            var list = new List<string>();
            foreach (var item in enumerable)
            {
                list.Add(item?.ToString() ?? string.Empty);
            }
            ConsoleHelpers.WriteDebugLine($"AsList: IEnumerable [{string.Join(", ", list)}]");
            return list;
        }

        // Default - empty list
        ConsoleHelpers.WriteDebugLine($"AsList: unknown type ({_value?.GetType().Name ?? "null"}) - returning empty list");
        return new List<string>();
    }

    public bool IsNotFound()
    {
        return Source == ConfigSource.NotFound;
    }

    public bool IsNotFoundNullOrEmpty()
    {
        if (Source == ConfigSource.NotFound)
        {
            return true;
        }

        if (_value == null)
        {
            return true;
        }

        if (_value is string stringValue)
        {
            return string.IsNullOrEmpty(stringValue);
        }

        if (_value is List<object> objList)
        {
            return objList.Count == 0;
        }

        if (_value is List<string> stringList)
        {
            return stringList.Count == 0;
        }

        return false;
    }

    public bool AddToList(string value)
    {
        var list = AsList();
        if (!list.Contains(value))
        {
            list.Add(value);
            _value = list;
            return true;
        }

        return false;
    }

    public bool RemoveFromList(string value)
    {
        var list = AsList();
        var removed = list.Remove(value);
        if (removed)
        {
            _value = list;
        }

        return removed;
    }

    public void Clear()
    {
        _value = null;
    }

    public void Set(object? value)
    {
        _value = value;
    }

    public override string ToString()
    {
        if (_value == null)
        {
            return string.Empty;
        }

        if (_value is List<object> objList)
        {
            return string.Join(", ", objList.Select(o => o?.ToString() ?? string.Empty));
        }

        if (_value is List<string> stringList)
        {
            return string.Join(", ", stringList);
        }

        return _value.ToString() ?? string.Empty;
    }

    private object? _value;
}

File: src/commonjs/Configuration/ConfigValue.ts
import { ConfigSource } from './ConfigSource';
import { ConfigFile } from './ConfigFile';

export class ConfigValue {
    public constructor();
    public constructor(value: any);
    public constructor(value: any, source: ConfigSource);
    public constructor(value: any, source: ConfigSource, isSecret: boolean);
    public constructor(value?: any, source?: ConfigSource, isSecret?: boolean) {
        this._value = value ?? null;
        this.Source = source ?? ConfigSource.NotFound;
        this.IsSecret = isSecret ?? false;
    }

    public get Value(): any { 
        return this._value; 
    }
    
    /**
     * Gets the source of this configuration value.
     */
    public readonly Source: ConfigSource;

    /**
     * Gets the configuration file associated with this value, if any.
     */
    public File: ConfigFile | null = null;

    /**
     * Gets whether this value should be treated as a secret.
     */
    public readonly IsSecret: boolean;

    public AsString(): string | null {
        return this._value?.toString() ?? null;
    }
    
    /**
     * Returns an obfuscated version of the value if it's a secret.
     * @returns The obfuscated value if it's a secret, otherwise the normal string representation.
     */
    public AsObfuscated(): string | null {
        if (!this.IsSecret || this._value == null) {
            return this.AsString();
        }
        
        const valueStr = this._value.toString() || '';
        if (valueStr.length <= 4) {
            return '****';
        }
        
        // Show first 2 and last 2 characters, obfuscate the rest
        return valueStr.substring(0, 2) + '*'.repeat(Math.min(60, valueStr.length - 4)) + valueStr.substring(valueStr.length - 2);
    }

    public AsInt(defaultValue: number = 0): number {
        if (typeof this._value === 'number') {
            return this._value;
        }

        if (typeof this._value === 'string') {
            const parsed = parseInt(this._value, 10);
            if (!isNaN(parsed)) {
                return parsed;
            }
        }

        return defaultValue;
    }

    public AsBool(defaultValue: boolean = false): boolean {
        if (typeof this._value === 'boolean') {
            return this._value;
        }

        if (typeof this._value === 'string') {
            const lower = this._value.toLowerCase();
            if (lower === 'true' || lower === 'false') {
                return lower === 'true';
            }
            return lower === 'true' || this._value === '1';
        }

        if (typeof this._value === 'number') {
            return this._value !== 0;
        }

        return defaultValue;
    }

    public AsList(): string[] {
        // Handle null value
        if (this._value == null) {
            console.debug('AsList: null value');
            return [];
        }

        // Handle arrays
        if (Array.isArray(this._value)) {
            const list = this._value.map(item => item?.toString() ?? '');
            console.debug(`AsList: Array [${list.join(', ')}]`);
            return list;
        }

        // Handle single string
        if (typeof this._value === 'string') {
            // Check if it's an empty array representation from YAML
            if (this._value === '[]') {
                console.debug('AsList: empty array string representation');
                return [];
            }
            
            console.debug(`AsList: single string ${this._value}`);
            return [this._value];
        }

        // Handle other iterable types (but not strings)
        if (this._value && typeof this._value[Symbol.iterator] === 'function' && typeof this._value !== 'string') {
            const list: string[] = [];
            for (const item of this._value) {
                list.push(item?.toString() ?? '');
            }
            console.debug(`AsList: Iterable [${list.join(', ')}]`);
            return list;
        }

        // Default - empty list
        console.debug(`AsList: unknown type (${typeof this._value}) - returning empty list`);
        return [];
    }

    public IsNotFound(): boolean {
        return this.Source === ConfigSource.NotFound;
    }

    public IsNotFoundNullOrEmpty(): boolean {
        if (this.Source === ConfigSource.NotFound) {
            return true;
        }

        if (this._value == null) {
            return true;
        }

        if (typeof this._value === 'string') {
            return this._value.length === 0;
        }

        if (Array.isArray(this._value)) {
            return this._value.length === 0;
        }

        return false;
    }

    public AddToList(value: string): boolean {
        const list = this.AsList();
        if (!list.includes(value)) {
            list.push(value);
            this._value = list;
            return true;
        }

        return false;
    }

    public RemoveFromList(value: string): boolean {
        const list = this.AsList();
        const index = list.indexOf(value);
        if (index > -1) {
            list.splice(index, 1);
            this._value = list;
            return true;
        }

        return false;
    }

    public Clear(): void {
        this._value = null;
    }

    public Set(value: any): void {
        this._value = value;
    }

    public toString(): string {
        if (this._value == null) {
            return '';
        }

        if (Array.isArray(this._value)) {
            return this._value.map(item => item?.toString() ?? '').join(', ');
        }

        return this._value.toString() ?? '';
    }

    private _value: any;
}